### ISSUE TITLE
multi-app.c example

### ISSUE THREAD

**Issue by bsomani** [curl/curl#4791]
### Some times (1 out of 10 times) when we run this multi-app example for downloading large file (700 MB) with FTP URL, the curl_multi_perform() API is blocked. When we took the backtrace using GDB, we see 

#0  0x00007fdb9ffb0fd0 in __write_nocancel () from /lib64/libc.so.6
#0  0x00007fdb9ffb0fd0 in __write_nocancel () from /lib64/libc.so.6
#1  0x00007fdb9ff3c0b3 in _IO_new_file_write () from /lib64/libc.so.6
#2  0x00007fdb9ff3c950 in __GI__IO_file_xsputn () from /lib64/libc.so.6
#3  0x00007fdb9ff315d2 in fwrite () from /lib64/libc.so.6
#4  0x00007fdba55ee672 in Curl_client_chop_write () from /lib64/libcurl.so.4
#5  0x00007fdba55e2fa4 in file_do () from /lib64/libcurl.so.4
#6  0x00007fdba560f366 in multi_runsingle () from /lib64/libcurl.so.4
#7  0x00007fdba560fd31 in curl_multi_perform () from /lib64/libcurl.so.4

sometimes we see it is blocked on fclose() call.

### I expected the following
Since curl_multi_perform() is non blocking, it should not block. As my application is single threaded, i have to kill the process whenever it is blocked for more than 30 seconds.
### curl/libcurl version 7.54.1

[curl -V output]
curl 7.54.1 (x86_64-redhat-linux-gnu) libcurl/7.54.1 OpenSSL/1.0.1e zlib/1.2.7 libidn2/2.0.4 libpsl/0.7.0 (+libicu/50.1.2) libssh2/1.4.3 nghttp2/1.21.1
Release-Date: 2017-06-14
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz HTTP2 UnixSockets HTTPS-proxy Metalink PSL 

### operating system
CentOS 7.6

**Comment by bagder** [comment]
That seems to be stuck within `fwrite()`, and thus outside of what libcurl controls. Could be a kernel or libc problem. File system?

**Comment by bsomani** [comment]
> That seems to be stuck within `fwrite()`, and thus outside of what libcurl controls. Could be a kernel or libc problem. File system?

Sorry, the earlier back strace (fwrite()) was found only once, but below backtrace is seen everytime whenever my application is stalled.
File system is ext4

#0  0x00007f6c126966e0 in __read_nocancel () from /lib64/libpthread.so.0
#1  0x00007f6c1220ff6d in file_do () from /lib64/libcurl.so.4
#2  0x00007f6c1223c366 in multi_runsingle () from /lib64/libcurl.so.4
#3  0x00007f6c1223cd31 in curl_multi_perform () from /lib64/libcurl.so.4
#4  0x0000000001c6ea9b in curlMultiPerform() ()

Basically this crash also happens when the source FTP url path is of local /tmp folder of same system.

**Comment by bsomani** [comment]
Basically curl_multi_perform() call is freezing my application.

**Comment by bagder** [comment]
Still, I don't see how libcurl can prevent this. Your system read function should not block like this when reading a file.

**Comment by bsomani** [comment]
ok, is there any way to provide timeout for libcurl API, so that even if it got blocked, then it will come out after that timeout expired.

**Comment by bagder** [comment]
**Your system read function should not block like this**. That's a bug, an issue, a problem. But it's not curl's problem, we just use the `read()` function. We don't have a way to "abort" a stuck read function because this doesn't happen. Or didn't happen. You're the first person ever to report this problem.

Additionally: file reading cannot be done non-blockingly in a portable fashion so a very large file transfer will actually block until done.

The regular libcurl timeouts work for file transfers though. They do not however fix the problem when `read()` unexpectedly get stuck for eons. That's not normal behavior.
